File: Compiler/CodeGeneration/Generator.py
from Translator.Model.Translation import Translation
from Translator.Model.Program import Program
import MemoryManager


class Generator:
	def __init__(self) -> None:
		self.memoryManager = None
		self.out = None

	'''
	Call to generate the program to file.
	'''
	def generate(self, program:Program, file:str) -> None:
		# Open file
		self.out = open(file, "w")
		
		# Recurse over instructions and write code
		self.memoryManager = MemoryManager(self.out)
		self.addGlobals(program.getGlobals())
		self.start(program.getStart())

		# Close file
		self.out.close()

	'''
	Call to add all global variables in memory
	'''
	def addGlobals(self, globals:set(Translation)):
		for variable in globals:
			self.memoryManager.addGlobal(variable)
	
	'''
	Call to start code traversal. This will traverse in DFS to print code for instructions.
	'''
	def start(self, start:Translation):
		pass

	'''
	Call to add a function.
	'''
	def addFunction(self, function):
		pass

	'''
	Call to a class.
	'''
	def addClass(self, definition):
		pass

	

File: src/main/java/compiler/Generator/Generator.java
package compiler.Generator;

import java.io.File;
import java.util.Set;

import compiler.Translator.Model.Program;
import compiler.Translator.Model.Translation;
import compiler.Translator.Model.Variable;
import compiler.Translator.Translations.*;
import compiler.Translator.Translations.Class;
import compiler.helpers.Printer;

public class Generator {
  private MemoryManager manager;
  private Printer printer;

  public Generator() {}

  public void generate(Program program, File file) {
    this.printer = new Printer(file);
    this.manager = new MemoryManager();
    this.addGlobal(program.getGlobal());
    this.start(program.getStart());
  }

  public void addGlobal(Set<Variable> globals) {
    this.manager.addGlobal(globals); 
  }

  public void start(Translation first) {
    Translation node = first;
    while (node.getNext() != null) {
      node.visit(this);
      node = node.getNext();
    }
  }

  public String visit(Assign node) {
    
    return null;
  }
  public String visit(Class node) {
    return null;
  }
  public String visit(For node) {
    return null;
  }
  public String visit(Function node) {
    return null;
  }
  public String visit(IfElse node) {
    return null;
  }
  public String visit(Operate node) {
    return null;
  }
  public String visit(While node) {
    return null;
  }
}
